### PR TITLE
Country flags in api

### PR DESCRIPTION
**Pull request by nickygerritsen**

### CODE REVIEW

**Comment by nickygerritsen** [comment]
For the country flags I regenerated them using the mentioned repository. Mauritania didn't have the correct flag, so I used my Inkscape and SVG edit skills to update the flag with the new red borders. Copied the color from the Wikipedia flag SVG.
Also updated codes.txt to reflect changes in country names where applicable.

Note that the CCS spec suggests:

> A server is recommended to provide flags of size around 56x56 and 160x160.

But I think it's not very useful to add 160x160 (ish) flags only to expoose them through the API. What do others think?
Our 80x60 flags are close enough to 56x56 IMHO.

**Comment by meisterT** [comment]
Are there actually tools that use flags from the API instead of bringing their own?

**Comment by nickygerritsen** [comment]
The ICPC tools as of https://github.com/icpctools/icpctools/commit/e55beb3cb9ed521473ce377686cf1d01053aa262
And the idea for the change in the CCS spec is that the API supports it and thus there is one place where you configure whether flags are available instead of in every tool.

File: webapp/src/Utils/Utils.php [hunk 1]
```diff
@@ -243,8 +243,10 @@ class Utils
         "FJI" => "Fiji",
         "FIN" => "Finland",
         "FRA" => "France",
-        "GUF" => "French Guiana",
+        // Removed since not an "official" country
+        // "GUF" => "French Guiana",
         "PYF" => "French Polynesia",
+        // Removed since not an "official" country
         "ATF" => "French Southern Territories",
```
Comment (meisterT) [inline webapp/src/Utils/Utils.php:250]: Did you mean to comment this line?

**Review by meisterT — APPROVED** [review]

File: webapp/src/Utils/Utils.php (hunk 1 above)
Comment (nickygerritsen) [reply webapp/src/Utils/Utils.php:250]: I did

**Comment by thijskh** [comment]
Maybe thinking a bit too much outside of the box, but what about something like https://github.com/lipis/flag-icon-css?
So we have nice SVGs that will work well in any format instead of PNGs in hardcoded dimensions?

**Comment by nickygerritsen** [comment]
I didn’t take something like this for two reasons:
* we now have “glossy” flags. Do we want to keep that? Can we with this solution?
* how do we expose them as png in the api? Can we?

**Comment by meisterT** [comment]
Although I find the flags that we currently have a little bit more visually appealing, this will also look fine, SVGs are great and the repository is kept up to date. So overall I like Thijs' idea.

**Comment by meisterT** [comment]
Ah, for the API: can we change the accepted formats there to include SVGs? It's likely better for downstream clients to have a high quality data source that they can convert to their needs instead of "arbitrary" resolution PNGs.

**Comment by nickygerritsen** [comment]
I will put this on the ccs spec agenda

**Comment by nickygerritsen** [comment]
I have updated this PR to use the SVG library suggested, which existed as a composer dependency so we can even easily keep it up to date.

Note that I am not using a `<span>` tag like they suggested, since that requires more CSS changes on our side to make it line up. [Any browser released in the last 10 years](https://caniuse.com/svg-img) supports rendering an SVG image inside an `` tag, so that seems like the easier solution.

The API now exposes both the square as well as the 4x3 flags.

I also got rid of our big countries array, since Symfony has a built in class that can do this for us.

**Comment by nickygerritsen** [comment]
Changed URLs: 
/public

File: doc/manual/config-advanced.rst [hunk 2]
```diff
@@ -10,10 +10,6 @@ You can place the images under the path `public/images/` (see
 the Config checker in the admin interfae for the full filesystem
 path of your installation) as follows:
 
-- *Country flags* are shown when the ``show_flags`` configuration option
```
Comment (meisterT) [inline doc/manual/config-advanced.rst:13]: why did you remove this?

File: webapp/src/Controller/API/GeneralInfoController.php [hunk 3]
```diff
@@ -286,6 +288,48 @@ public function getConfigCheckAction(): JsonResponse
...
+     */
+    public function countryFlagAction(string $countryCode, string $size): Response
+    {
+        // This API action exists for two reasons
+        // - Relative URL's are relative to the API root according to the CCS spec. This
+        //   means that we need to have an API endpoint for files.
+        // - This makes it that we can not return a flag if flags are disabled.
+
+        if (!$this->config->get('show_flags')) {
+            throw new NotFoundHttpException('country flag not found');
```
Comment (meisterT) [inline webapp/src/Controller/API/GeneralInfoController.php:314]: Well, we might have found it, but it's not enabled :-)

File: webapp/src/Controller/API/GeneralInfoController.php [hunk 4]
```diff
@@ -286,6 +288,48 @@ public function getConfigCheckAction(): JsonResponse
...
+        }
+
+        $alpha3code = strtoupper($countryCode);
+        if (!Countries::alpha3CodeExists($alpha3code)) {
+            throw new NotFoundHttpException('country flag not found');
```
Comment (meisterT) [inline webapp/src/Controller/API/GeneralInfoController.php:319]: mention the missing country code in the message?

File: webapp/src/Controller/API/GeneralInfoController.php [hunk 5]
```diff
@@ -286,6 +288,48 @@ public function getConfigCheckAction(): JsonResponse
...
+        }
+        $alpha2code = strtolower(Countries::getAlpha2Code($alpha3code));
+        $flagFile = sprintf('%s/public/flags/%s/%s.svg', $this->dj->getDomjudgeWebappDir(), $size, $alpha2code);
+
+        if (!file_exists($flagFile)) {
+            throw new NotFoundHttpException('country flag not found');
```
Comment (meisterT) [inline webapp/src/Controller/API/GeneralInfoController.php:325]: How is this different from the one above?

File: webapp/src/DataFixtures/Test/SampleAffilicationsFixture.php [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+use Doctrine\Bundle\FixturesBundle\Fixture;
+use Doctrine\Persistence\ObjectManager;
+
+class SampleAffilicationsFixture extends Fixture
+{
+    public function load(ObjectManager $manager)
+    {
+        $affiliationData = [
+            // short name, name,                                                country
+            ['FAU',        'Friedrich-Alexander-Universität Erlangen-Nürnberg', 'DEU'],
```
Comment (meisterT) [inline webapp/src/DataFixtures/Test/SampleAffilicationsFixture.php:15]: :heart_eyes:

File: doc/manual/config-advanced.rst (hunk 2 above)
Comment (nickygerritsen) [reply doc/manual/config-advanced.rst:13]: Whoops

File: webapp/src/Controller/API/GeneralInfoController.php (hunk 3 above)
Comment (nickygerritsen) [reply webapp/src/Controller/API/GeneralInfoController.php:314]: Well I didn’t want to expose this difference but if you think we should I can change it

File: webapp/src/Controller/API/GeneralInfoController.php (hunk 5 above)
Comment (nickygerritsen) [reply webapp/src/Controller/API/GeneralInfoController.php:325]: If we somehow not have an SVG for a valid alpha3/2 code

File: webapp/src/Controller/API/GeneralInfoController.php (hunk 3 above)
Comment (meisterT) [reply webapp/src/Controller/API/GeneralInfoController.php:314]: Yep, I think we should make this explicit to ease debugging.

**Comment by nickygerritsen** [comment]
Changed URLs:
/api/doc
/jury/affiliations
/jury/affiliations/1
/jury/affiliations/1/edit
/jury/affiliations/add
/jury
/jury/scoreboard
/style/.*
/team/scoreboard
/team/team/2

**Review by thijskh — APPROVED** [review]
